core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java

```diff
@@ -141,51 +141,57 @@ public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
-        } else if (nodeDecision.type() == Decision.Type.NO) {
-            finalDecision = ClusterAllocationExplanation.FinalDecision.NO;
-            finalExplanation = "the shard cannot be assigned because one or more allocation decider returns a 'NO' decision";
+        if (storeStatus == null) {
+            // No copies of the data
+            storeCopy = ClusterAllocationExplanation.StoreCopy.NONE;
         } else {
-            finalDecision = ClusterAllocationExplanation.FinalDecision.YES;
-            finalExplanation = "the shard can be assigned";
-        }
-
-        if (storeStatus != null) {
             final Throwable storeErr = storeStatus.getStoreException();
-            // The store error only influences the decision if the shard is primary and has not been allocated before
-            if (storeErr != null && shard.primary() && shard.allocatedPostIndexCreate(indexMetaData) == false) {
-                finalDecision = ClusterAllocationExplanation.FinalDecision.NO;
+            if (storeErr != null) {
                 if (ExceptionsHelper.unwrapCause(storeErr) instanceof CorruptIndexException) {
                     storeCopy = ClusterAllocationExplanation.StoreCopy.CORRUPT;
-                    finalExplanation = "the copy of data in the shard store is corrupt";
                 } else {
                     storeCopy = ClusterAllocationExplanation.StoreCopy.IO_ERROR;
-                    finalExplanation = "there was an IO error reading from data in the shard store";
                 }
             } else if (activeAllocationIds.isEmpty()) {
                 // The ids are only empty if dealing with a legacy index
                 // TODO: fetch the shard state versions and display here?
                 storeCopy = ClusterAllocationExplanation.StoreCopy.UNKNOWN;
             } else if (activeAllocationIds.contains(storeStatus.getAllocationId())) {
                 storeCopy = ClusterAllocationExplanation.StoreCopy.AVAILABLE;
-                if (finalDecision == ClusterAllocationExplanation.FinalDecision.YES) {
-                    finalExplanation = "the shard can be assigned and the node contains a valid copy of the shard data";
-                }
             } else {
                 // Otherwise, this is a stale copy of the data (allocation ids don't match)
                 storeCopy = ClusterAllocationExplanation.StoreCopy.STALE;
-                finalExplanation = "the copy of the shard is stale, allocation ids do not match";
-                finalDecision = ClusterAllocationExplanation.FinalDecision.NO;
             }
+        }
+
+        if (node.getId().equals(assignedNodeId)) {
+            finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
+            finalExplanation = "the shard is already assigned to this node";
+        } else if (shard.primary() && shard.unassigned() && storeCopy == ClusterAllocationExplanation.StoreCopy.STALE) {
```

Comment:
My (hopefully) last corrections here (in this case I'm correcting my earlier self 😞): 
- We should also add allocatedPostIndexCreate() to be more in line with PrimaryShardAllocator. This means that this condition and the subsequent two become `shard.primary() && shard.unassigned() && shard.allocatedPostIndexCreate(indexMetaData) && storeCopy == ClusterAllocationExplanation.StoreCopy.STALE`

- We should also handle `StoreCopy.NONE` in the same way. In case of assigning an unassigned primary, having no shard data is also a NO.